File: MemoryGame/Services/HapticManager.swift
//
//  HapticManager.swift
//  Memory Match Kids
//
//  Semantic haptic feedback. Each event maps to the most fitting Taptic pattern,
//  and generators are pre-warmed via `prepare()` to minimise first-fire latency.
//  Every call takes the player's `enabled` preference so haptics stay off when
//  the setting is disabled.
//

import UIKit

enum HapticManager {
    private static let lightImpact = UIImpactFeedbackGenerator(style: .light)
    private static let mediumImpact = UIImpactFeedbackGenerator(style: .medium)
    private static let rigidImpact = UIImpactFeedbackGenerator(style: .rigid)
    private static let notification = UINotificationFeedbackGenerator()
    private static let selectionGen = UISelectionFeedbackGenerator()

    /// Warm up the engines at the start of a game to cut first-tap latency.
    static func prepare() {
        lightImpact.prepare()
        notification.prepare()
    }

    // MARK: - Semantic events

    /// A card is flipped face-up.
    static func cardFlip(enabled: Bool) {
        guard enabled else { return }
        lightImpact.impactOccurred(intensity: 0.7)
        lightImpact.prepare()
    }

    /// A pair / group matched.
    static func match(enabled: Bool) {
        guard enabled else { return }
        notification.notificationOccurred(.success)
        notification.prepare()
    }

    /// A wrong guess.
    static func mismatch(enabled: Bool) {
        guard enabled else { return }
        notification.notificationOccurred(.error)
    }

    /// A life / heart was lost.
    static func lifeLost(enabled: Bool) {
        guard enabled else { return }
        rigidImpact.impactOccurred()
    }

    /// Level finished with a win.
    static func levelComplete(enabled: Bool) {
        guard enabled else { return }
        notification.notificationOccurred(.success)
    }

    /// Level lost (out of moves / time / lives).
    static func levelFailed(enabled: Bool) {
        guard enabled else { return }
        notification.notificationOccurred(.warning)
    }

    /// A new achievement unlocked.
    static func achievement(enabled: Bool) {
        guard enabled else { return }
        notification.notificationOccurred(.success)
    }

    /// Light selection tick (toggles, segmented choices, theme change).
    static func selection(enabled: Bool) {
        guard enabled else { return }
        selectionGen.selectionChanged()
    }

    // MARK: - Legacy generic names (retained for any older call sites)

    static func light(enabled: Bool) { cardFlip(enabled: enabled) }
    static func success(enabled: Bool) { match(enabled: enabled) }
    static func error(enabled: Bool) { mismatch(enabled: enabled) }
    static func medium(enabled: Bool) {
        guard enabled else { return }
        mediumImpact.impactOccurred()
    }
}

File: android/app/src/main/java/com/memogame/app/services/HapticManager.kt
package com.memogame.app.services

import android.content.Context
import android.os.Build
import android.os.VibrationEffect
import android.os.Vibrator
import android.os.VibratorManager

/**
 * Semantic haptic feedback — port of the iOS HapticManager. Every call takes
 * the player's `enabled` preference so haptics stay off when disabled.
 */
object HapticManager {
    private var vibrator: Vibrator? = null

    fun init(context: Context) {
        vibrator = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.S) {
            val manager = context.getSystemService(Context.VIBRATOR_MANAGER_SERVICE) as? VibratorManager
            manager?.defaultVibrator
        } else {
            @Suppress("DEPRECATION")
            context.getSystemService(Context.VIBRATOR_SERVICE) as? Vibrator
        }
    }

    private fun vibrate(enabled: Boolean, durationMs: Long, amplitude: Int) {
        if (!enabled) return
        val v = vibrator ?: return
        runCatching {
            v.vibrate(VibrationEffect.createOneShot(durationMs, amplitude))
        }
    }

    private fun pattern(enabled: Boolean, timings: LongArray, amplitudes: IntArray) {
        if (!enabled) return
        val v = vibrator ?: return
        runCatching {
            v.vibrate(VibrationEffect.createWaveform(timings, amplitudes, -1))
        }
    }

    /** A card is flipped face-up. */
    fun cardFlip(enabled: Boolean) = vibrate(enabled, 18, 120)

    /** A pair / group matched. */
    fun match(enabled: Boolean) = pattern(enabled, longArrayOf(0, 25, 60, 35), intArrayOf(0, 140, 0, 200))

    /** A wrong guess. */
    fun mismatch(enabled: Boolean) = pattern(enabled, longArrayOf(0, 40, 60, 40), intArrayOf(0, 180, 0, 180))

    /** A life / heart was lost. */
    fun lifeLost(enabled: Boolean) = vibrate(enabled, 45, 255)

    /** Level finished with a win. */
    fun levelComplete(enabled: Boolean) =
        pattern(enabled, longArrayOf(0, 30, 60, 30, 60, 60), intArrayOf(0, 140, 0, 180, 0, 255))

    /** Level lost (out of moves / time / lives). */
    fun levelFailed(enabled: Boolean) = pattern(enabled, longArrayOf(0, 60, 80, 60), intArrayOf(0, 160, 0, 160))

    /** A new achievement unlocked. */
    fun achievement(enabled: Boolean) = match(enabled)

    /** Light selection tick (toggles, segmented choices, theme change). */
    fun selection(enabled: Boolean) = vibrate(enabled, 12, 80)
}
